README.md
# Baum v2

[![Build Status](https://travis-ci.org/gazsp/baum.svg?branch=2.0.1)](https://travis-ci.org/gazsp/baum)
[![Coverage Status](https://coveralls.io/repos/github/gazsp/baum/badge.svg?branch=2.0.1)](https://coveralls.io/github/gazsp/baum?branch=2.0.1)
[![StyleCI](https://github.styleci.io/repos/47506280/shield?branch=2.0.1&style=flat)](https://github.styleci.io/repos/47506280)

## Nested Set implementation for Laravel

Baum is an implementation of the [Nested Set](http://en.wikipedia.org/wiki/Nested_set_model)
pattern for the [Laravel](http://laravel.com/) Eloquent ORM.

---

[![this](https://img.shields.io/badge/warning-work%20in%20progress-red)]()

*If you find a bug*, **please** file an issue and submit a pull request with a failing unit test.

---

## Installation

Baum v2 works with Laravel 5.8 (and PHP 7.2) and above. You can add it to your project with:
```
composer require gazsp/baum
```
## Documentation

* [About Nested Sets](#about)
* [The theory behind, a TL;DR version](#theory)
* [Installation](#installation)
* [Getting started](#getting-started)
* [Usage](#usage)
* [Further information](#further-information)
* [Contributing](#contributing)

<a name="about"></a>
## About Nested Sets

A nested set is a smart way to implement an _ordered_ tree that allows for fast,
non-recursive queries. For example, you can fetch all descendants of a node in a
single query, no matter how deep the tree. The drawback is that insertions/moves/deletes
require complex SQL, but that is handled behind the curtains by this package!

Nested sets are appropriate for ordered trees (e.g. menus, commercial categories)
and big trees that must be queried efficiently (e.g. threaded posts).

See the [wikipedia entry for nested sets](http://en.wikipedia.org/wiki/Nested_set_model)
for more info. Also, this is a good introductory tutorial:
[http://www.evanpetersen.com/item/nested-sets.html](http://www.evanpetersen.com/item/nested-sets.html)

<a name="theory"></a>
## The theory behind, a TL;DR version

An easy way to visualize how a nested set works is to think of a parent entity surrounding all
of its children, and its parent surrounding it, etc. So this tree:

    root
      |_ Child 1
        |_ Child 1.1
        |_ Child 1.2
      |_ Child 2
        |_ Child 2.1
        |_ Child 2.2


Could be visualized like this:

     ___________________________________________________________________
    |  Root                                                             |
    |    ____________________________    ____________________________   |
    |   |  Child 1                  |   |  Child 2                  |   |
    |   |   __________   _________  |   |   __________   _________  |   |
    |   |  |  C 1.1  |  |  C 1.2 |  |   |  |  C 2.1  |  |  C 2.2 |  |   |
    1   2  3_________4  5________6  7   8  9_________10 11_______12 13  14
    |   |___________________________|   |___________________________|   |
    |___________________________________________________________________|

The numbers represent the left and right boundaries.  The table then might
look like this:

    id | parent_id | lft  | rgt  | depth | data
     1 |           |    1 |   14 |     0 | root
     2 |         1 |    2 |    7 |     1 | Child 1
     3 |         2 |    3 |    4 |     2 | Child 1.1
     4 |         2 |    5 |    6 |     2 | Child 1.2
     5 |         1 |    8 |   13 |     1 | Child 2
     6 |         5 |    9 |   10 |     2 | Child 2.1
     7 |         5 |   11 |   12 |     2 | Child 2.2

To get all children of a _parent_ node, you

```sql
SELECT * WHERE lft IS BETWEEN parent.lft AND parent.rgt
```

To get the number of children, it's

```sql
(right - left - 1)/2
```

To get a node and all its ancestors going back to the root, you

```sql
SELECT * WHERE node.lft IS BETWEEN lft AND rgt
```

As you can see, queries that would be recursive and prohibitively slow on
ordinary trees are suddenly quite fast. Nifty, isn't it?

<a name="installation"></a>
## Installation

Baum works with Laravel 5.5 (and PHP 7) and above. You can add it to your `composer.json` file
with:

    "gazsp/baum": "2.0"

Run `composer install` to install it.

As with most Laravel 5 packages you'll then need to register the Baum
*service provider*. To do that, head over your `config/app.php` file and add
the following line into the `providers` array:

    'Baum\Providers\BaumServiceProvider',

<a name="getting-started"></a>
## Getting started

After the package is correctly installed the easiest way to get started is to
run the provided generator:

    php artisan baum:install MODEL

Replace model by the class name you plan to use for your Nested Set model.

The generator will install a migration and a model file into your application
configured to work with the Nested Set behaviour provided by Baum. You SHOULD
take a look at those files, as each of them describes how they can be customized.

Next, you would probably run `artisan migrate` to apply the migration.

### Model configuration

In order to work with Baum, you must ensure that your model class extends
`Baum\Node`.

This is the easiest it can get:

```php
class Category extends Baum\Node {

}
```

This is a *slightly* more complex example where we have the column names customized:

```php
class Dictionary extends Baum\Node {

  protected $table = 'dictionary';

  // 'parent_id' column name
  protected $parentColumn = 'parent_id';

  // 'lft' column name
  protected $leftColumn = 'lidx';

  // 'rgt' column name
  protected $rightColumn = 'ridx';

  // 'depth' column name
  protected $depthColumn = 'nesting';

  // guard attributes from mass-assignment
  protected $guarded = array('id', 'parent_id', 'lidx', 'ridx', 'nesting');

}
```

Remember that, obviously, the column names must match those in the database table.

### Migration configuration

You must ensure that the database table that supports your Baum models has the
following columns:

* `parent_id`: a reference to the parent (int)
* `lft`: left index bound (int)
* `rgt`: right index bound (int)
* `depth`: depth or nesting level (int)

Here is a sample migration file:

```php
class Category extends Migration {

  public function up() {
    Schema::create('categories', function(Blueprint $table) {
      $table->increments('id');

      $table->integer('parent_id')->nullable();
      $table->integer('lft')->nullable();
      $table->integer('rgt')->nullable();
      $table->integer('depth')->nullable();

      $table->string('name', 255);

      $table->timestamps();
    });
  }

  public function down() {
    Schema::drop('categories');
  }

}
```

You may freely modify the column names, provided you change them both in the
migration and the model.

<a name="usage"></a>
## Usage

After you've configured your model and run the migration, you are now ready
to use Baum with your model. Below are some examples.

* [Creating a root node](#creating-root-node)
* [Inserting nodes](#inserting-nodes)
* [Deleting nodes](#deleting-nodes)
* [Getting the nesting level of a node](#node-level)
* [Moving nodes around](#moving-nodes)
* [Asking questions to your nodes](#node-questions)
* [Relations](#node-relations)
* [Root and Leaf scopes](#node-basic-scopes)
* [Accessing the ancestry/descendancy chain](#node-chains)
* [Limiting levels of children returned](#limiting-depth)
* [Custom sorting column](#custom-sorting-column)
* [Dumping the hierarchy tree](#hierarchy-tree)
* [Model events: `moving` and `moved`](#node-model-events)
* [Scope support](#scope-support)
* [Validation](#validation)
* [Tree rebuilding](#rebuilding)
* [Soft deletes](#soft-deletes)
* [Seeding/Mass assignment](#seeding)
* [Misc/Utility functions](#misc-utilities)

<a name="creating-root-node"></a>
### Creating a root node

By default, all nodes are created as roots:

```php
$root = Category::create(['name' => 'Root category']);
```

Alternatively, you may find yourself in the need of *converting* an existing node
into a *root node*:

```php
$node->makeRoot();
```

You may also nullify it's `parent_id` column to accomplish the same behaviour:

```php
// This works the same as makeRoot()
$node->parent_id = null;
$node->save();
```

<a name="inserting-nodes"></a>
### Inserting nodes

```php
// Directly with a relation
$child1 = $root->children()->create(['name' => 'Child 1']);

// with the `makeChildOf` method
$child2 = Category::create(['name' => 'Child 2']);
$child2->makeChildOf($root);
```

<a name="deleting-nodes"></a>
### Deleting nodes

```php
$child1->delete();
```

Descendants of deleted nodes will also be deleted and all the `lft` and `rgt`
bound will be recalculated. Pleases note that, for now, `deleting` and `deleted`
model events for the descendants will not be fired.

<a name="node-level"></a>
### Getting the nesting level of a node

The `getLevel()` method will return current nesting level, or depth, of a node.

```php
$node->getLevel() // 0 when root
```

<a name="moving-nodes"></a>
### Moving nodes around

Baum provides several methods for moving nodes around:

* `moveLeft()`: Find the left sibling and move to the left of it.
* `moveRight()`: Find the right sibling and move to the right of it.
* `moveToLeftOf($otherNode)`: Move to the node to the left of ...
* `moveToRightOf($otherNode)`: Move to the node to the right of ...
* `makeNextSiblingOf($otherNode)`: Alias for `moveToRightOf`.
* `makeSiblingOf($otherNode)`: Alias for `makeNextSiblingOf`.
* `makePreviousSiblingOf($otherNode)`: Alias for `moveToLeftOf`.
* `makeChildOf($otherNode)`: Make the node a child of ...
* `makeFirstChildOf($otherNode)`: Make the node the first child of ...
* `makeLastChildOf($otherNode)`: Alias for `makeChildOf`.
* `makeRoot()`: Make current node a root node.

For example:

```php
$root = Creatures::create(['name' => 'The Root of All Evil']);

$dragons = Creatures::create(['name' => 'Here Be Dragons']);
$dragons->makeChildOf($root);

$monsters = new Creatures(['name' => 'Horrible Monsters']);
$monsters->save();

$monsters->makeSiblingOf($dragons);

$demons = Creatures::where('name', '=', 'demons')->first();
$demons->moveToLeftOf($dragons);
```

<a name="node-questions"></a>
### Asking questions to your nodes

You can ask some questions to your Baum nodes:

* `isRoot()`: Returns true if this is a root node.
* `isLeaf()`: Returns true if this is a leaf node (end of a branch).
* `isChild()`: Returns true if this is a child node.
* `isChildOf($other)`: Returns true if this node is a child of the other.
* `isDescendantOf($other)`: Returns true if node is a descendant of the other.
* `isSelfOrDescendantOf($other)`: Returns true if node is self or a descendant.
* `isAncestorOf($other)`: Returns true if node is an ancestor of the other.
* `isSelfOrAncestorOf($other)`: Returns true if node is self or an ancestor.
* `equals($node)`: current node instance equals the other.
* `insideSubtree($node)`: Checks whether the given node is inside the subtree
defined by the left and right indices.
* `inSameScope($node)`: Returns true if the given node is in the same scope
as the current one. That is, if *every* column in the `scoped` property has
the same value in both nodes.

Using the nodes from the previous example:

```php
$demons->isRoot(); // => false

$demons->isDescendantOf($root) // => true
```

<a name="node-relations"></a>
### Relations

Baum provides two self-referential Eloquent relations for your nodes: `parent`
and `children`.

```php
$parent = $node->parent()->get();

$children = $node->children()->get();
```

<a name="node-basic-scopes"></a>
### Root and Leaf scopes

Baum provides some very basic query scopes for accessing the root and leaf nodes:

```php
// Query scope which targets all root nodes
Category::roots()

// All leaf nodes (nodes at the end of a branch)
Category:allLeaves()
```

You may also be interested in only the first root:

```php
$firstRootNode = Category::root();
```

<a name="node-chains"></a>
### Accessing the ancestry/descendancy chain

There are several methods which Baum offers to access the ancestry/descendancy
chain of a node in the Nested Set tree. The main thing to keep in mind is that
they are provided in two ways:

First as **query scopes**, returning an `Illuminate\Database\Eloquent\Builder`
instance to continue to query further. To get *actual* results from these,
remember to call `get()` or `first()`.

* `ancestorsAndSelf()`: Targets all the ancestor chain nodes including the current one.
* `ancestors()`: Query the ancestor chain nodes excluding the current one.
* `siblingsAndSelf()`: Instance scope which targets all children of the parent, including self.
* `siblings()`: Instance scope targeting all children of the parent, except self.
* `leaves()`: Instance scope targeting all of its nested children which do not have children.
* `descendantsAndSelf()`: Scope targeting itself and all of its nested children.
* `descendants()`: Set of all children & nested children.
* `immediateDescendants()`: Set of all children nodes (non-recursive).

Second, as **methods** which return actual `Baum\Node` instances (inside a `Collection`
object where appropiate):

* `getRoot()`: Returns the root node starting at the current node.
* `getAncestorsAndSelf()`: Retrieve all of the ancestor chain including the current node.
* `getAncestorsAndSelfWithoutRoot()`: All ancestors (including the current node) except the root node.
* `getAncestors()`: Get all of the ancestor chain from the database excluding the current node.
* `getAncestorsWithoutRoot()`: All ancestors except the current node and the root node.
* `getSiblingsAndSelf()`: Get all children of the parent, including self.
* `getSiblings()`: Return all children of the parent, except self.
* `getLeaves()`: Return all of its nested children which do not have children.
* `getDescendantsAndSelf()`: Retrieve all nested children and self.
* `getDescendants()`: Retrieve all of its children & nested children.
* `getImmediateDescendants()`: Retrieve all of its children nodes (non-recursive).

Here's a simple example for iterating a node's descendants (provided a name
attribute is available):

```php
$node = Category::where('name', '=', 'Books')->first();

foreach($node->getDescendantsAndSelf() as $descendant) {
  echo "{$descendant->name}";
}
```

<a name="limiting-depth"></a>
### Limiting the levels of children returned

In some situations where the hierarchy depth is huge it might be desirable to limit the number of levels of children returned (depth). You can do this in Baum by using the `limitDepth` query scope.

The following snippet will get the current node's descendants up to a maximum
of 5 depth levels below it:

```php
$node->descendants()->limitDepth(5)->get();
```

Similarly, you can limit the descendancy levels with both the `getDescendants` and `getDescendantsAndSelf` methods by supplying the desired depth limit as the first argument:

```php
// This will work without depth limiting
// 1. As usual
$node->getDescendants();
// 2. Selecting only some attributes
$other->getDescendants(array('id', 'parent_id', 'name'));
...
// With depth limiting
// 1. A maximum of 5 levels of children will be returned
$node->getDescendants(5);
// 2. A max. of 5 levels of children will be returned selecting only some attrs
$other->getDescendants(5, array('id', 'parent_id', 'name'));
```

<a name="custom-sorting-column"></a>
### Custom sorting column

By default in Baum all results are returned sorted by the `lft` index column
value for consistency.

If you wish to change this default behaviour you need to specify in your model
the name of the column you wish to use to sort your results like this:

```php
protected $orderColumn = 'name';
```

<a name="hierarchy-tree"></a>
### Dumping the hierarchy tree

Baum extends the default `Eloquent\Collection` class and provides the
`toHierarchy` method to it which returns a nested collection representing the
queried tree.

Retrieving a complete tree hierarchy into a regular `Collection` object with
its children *properly nested* is as simple as:

```php
$tree = Category::where('name', '=', 'Books')->first()->getDescendantsAndSelf()->toHierarchy();
```

<a name="node-model-events"></a>
### Model events: `moving` and `moved`

Baum models fire the following events: `moving` and `moved` every time a node
is *moved* around the Nested Set tree. This allows you to hook into those points
in the node movement process. As with normal Eloquent model events, if `false`
is returned from the `moving` event, the movement operation will be cancelled.

The recommended way to hook into those events is by using the model's boot
method:

```php
class Category extends Baum\Node {

  public static function boot() {
    parent::boot();

    static::moving(function($node) {
      // Before moving the node this function will be called.
    });

    static::moved(function($node) {
      // After the move operation is processed this function will be
      // called.
    });
  }

}
```

<a name="scope-support"></a>
### Scope support

Baum provides a simple method to provide Nested Set "scoping" which restricts
what we consider part of a nested set tree. This should allow for multiple nested
set trees in the same database table.

To make use of the scoping funcionality you may override the `scoped` model
attribute in your subclass. This attribute should contain an array of the column
names (database fields) which shall be used to restrict Nested Set queries:

```php
class Category extends Baum\Node {
  ...
  protected $scoped = array('company_id');
  ...
}
```

In the previous example, `company_id` effectively restricts (or "scopes") a
Nested Set tree. So, for each value of that field we may be able to construct
a full different tree.

```php
$root1 = Category::create(['name' => 'R1', 'company_id' => 1]);
$root2 = Category::create(['name' => 'R2', 'company_id' => 2]);

$child1 = Category::create(['name' => 'C1', 'company_id' => 1]);

CHANGELOG.md
## 2.0.1
_May 1 2020_

* Reworking unit tests.
* PHP 7.1 Deprecated
* Laravel 5.5 Deprecated - Due to dependency issue with carbon and PHP7.1

## 2.0.0
_Mar 16 2020_

* Starting work to move feature/2.0 to master.

## 1.1.5
_Mar 27 2015_

* Fixed bug with Install command in Lumen [https://github.com/gazsp/baum/issues/12](https://github.com/gazsp/baum/issues/12)
* Added getOthersAtSameDepth() method [https://github.com/gazsp/baum/commit/97dced5a6e40ed4437101aa2d53e0346cc1843f8](https://github.com/gazsp/baum/commit/97dced5a6e40ed4437101aa2d53e0346cc1843f8)

## 1.1.4
_Jan 5 2015_
* Updated to use latest Laravel 5.2 packages
* Now requires PHP >= v5.6.x

## 1.1.3
_Dec 27 2015_
* API changes
  - Removed 'force' parameter from `Node->rebuild`
  - Removed unused 'excludeDeleted' parameter from `Node->newNestedSetQuery`
* Bug fixes
  - `Node::all` now uses columns parameter in get call
* Documentation changes
  - Soft delete now not advised

## 1.1.2
* ???

## 1.1.1
_Mar 4 2015_
* bug fixes
  - Fix README.md typos [PR #134 and #131]
  - Use table prefixes in `rebuild` command [PR #133]
  - Fix `limitDepth` query scope [PR #125]

## 1.1.0
_Mar 3 2015_
* enhancements
  - Add support for Laravel 5.x, leave a [1.0.x](https://github.com/etrepat/baum/tree/1.0.x-stable) branch for backwards compatibility.

## 1.0.13
_Sep 30 2014_
* bug fixes
  - Fix node depth recomputation when moving nodes. Closes [#109] & [#106].
  - Laravel 5, composer compatibility.

## 1.0.12
_Aug 25 2014_
* bug fixes
  - Fix check for `Node->storeNewParent()` to avoid moving a node which was already root. Fixes [#79] and closes [#81].
  - Several git fixes by @GrahamCampbell [PR #96].
  - `.travis.yml` improvements by @GrahamCampbell [PR #95].
  - Rename `isValid` static method to `isValidNestedSet` to avoid name clashes with validator packages. Closes [#89].
  - Moving to root has been now reimplemented to be more consistent.

* enhancements
  - Add pessimistic locking into the library. Closes [#89].
  - Add `trunks` family of methods: `allTrunks`, `trunks`, `getTrunks` and `isTrunk`. Closes [#59].
  - Add a `force` option to the `rebuild` static method.

## 1.0.11
_Jul 3 2014_
* bug fixes
  - Target Laravel stable version in `composer.json` file. Merges [#67].
  - Use qualified order column name on `newNestedSetQuery` method. Merges [#70].
  - Cleaned up migration stub. Merges [#84].
  - Enforce sorting for `$orderColumn` when calling `toHierarchy`. Merges [#73], fixing [#71].

* enhancements
  - [#77] Add `makeFirstChildOf` and `makeLastChildOf` helper methods.
  - [#62] Implement `limitDepth` query scope to allow query depth limiting for huge descendancy chains. Also allow to pass the depth limit as the first parameter of `getDescendants` and `getDescendantsAndSelf`.
  - [#67] Should work with Laravel 4.2
  - [#68] Implement `buildTree`, `makeTree` mass-assignment (seeding) methods.

## 1.0.10
_May 2 2014_
* bug fixes
  - Fix inserting a new model with a defined scope. Fixes [#27].
  - Static methods now do not take into account scoped column values, which
  would not make sense.
  - Properly set a model's relations when reloading an instance via `reload()`.
  - Fix `getObservableEvents()` function to include node's `moving` and `moved`
  events.
  - Fix `reload()` to consider trashed node objects via soft-delete operations.
  Fixes [#35].
  - Preliminar support for soft-delete operations. Should fix [#23].
  - Assigning `null` to the `parent_id` column value and saving the node now
  performs the same operation as `makeRoot()`. Fixes [#54].
  - Reimplement `toHierarchy` as yielded inconsistent results, even worse with custom sorting of the collection results. Merges [#61], fixes [#63].

* enhancements
  - Implement tree *structure validation* via `Node::isValid`.
  - Implement tree *index rebuilding* via `Node::rebuild`. Very useful when a tree
  structure has been messed up or to convert from a `parent_id` only table.
  - Preliminar support for soft-delete operations.
  - Allow the user to change the default sorting column name (defaulting to `lft`).
  - Add support for non-numeric keys. Merges & fixes [#52].


## 1.0.9
_Jan 14 2014_

* bug fixes
  - [#26] Prevent impossible moves to the left or right. Now moving a node too
  further to the left or right raises a `MoveNotPossibleException`. Thanks to
  @ziadoz for spotting this issue and providing a patch.

* enhancenments
  - [#5] Implement `toHierarchy` method which returns a nested collection
  representing the queried tree. Great thanks go to @Surt for his patch on this.
  - Add a static `all` method which works as the regular `Eloquent\Model::all`
  method but sorts for the `lft` column.

## 1.0.8
_Oct 11 2013_

* bug fixes
  - Properly wrap column names when used inside raw queries.
  - Correct examples from README.

## 1.0.6
_July 31 2013_

* enhancements
  - Implement `withoutNode`, `withoutRoot`, `withoutSelf` query scopes.
  - [#15] Add `Node::getNestedList` static method thanks to @gerp.

## 1.0.5
_July 22 2013_

* enhancements
  - Implement simple means to implement "scopes" for a Nested Set tree, allowing
  for multiple trees in the same database table.

## 1.0.4
_July 18 2013_

* bug fixes
  - Full support for PHP >= 5.3.7. Baum should work if Eloquent works.

## 1.0.3
_July 5th 2013_

* enhanments
  - Add support for PHP 5.3 to make Baum match the requirements of Eloquent.
  - Improve Postgres friendliness

* bug fixes
  - `insideSubtree` was inside incorrectly to guard against impossible moves. Fixed.
  - Due to default ordering by `lft` column, Postgres complained when using
  aggregate methods on queries which had SORT BY clauses. Now those are pruned
  before running aggregates.


## 1.0.2
_July 3rd 2013_

* enhancements
  - Add `immediateDescendants`, 'getImmediateDescendants' methods.
  - Implement a test suite.

* bug fixes
  - [#8] Add `lft` logic ordering by default to every scope method. Thanks to
  @dirkpostma for spotting this out.

.tests/seeders/CategorySeeder.php
<?php

class CategorySeeder
{
    public function run()
    {
        Category::unguard();

        Category::create(['id' => 1, 'name' => 'Root 1', 'lft' => 1, 'rgt' => 10, 'depth' => 0]);
        Category::create(['id' => 2, 'name' => 'Child 1', 'lft' => 2, 'rgt' => 3, 'depth' => 1, 'parent_id' => 1]);
        Category::create(['id' => 3, 'name' => 'Child 2', 'lft' => 4, 'rgt' => 7, 'depth' => 1, 'parent_id' => 1]);
        Category::create(['id' => 4, 'name' => 'Child 2.1', 'lft' => 5, 'rgt' => 6, 'depth' => 2, 'parent_id' => 3]);
        Category::create(['id' => 5, 'name' => 'Child 3', 'lft' => 8, 'rgt' => 9, 'depth' => 1, 'parent_id' => 1]);
        Category::create(['id' => 6, 'name' => 'Root 2', 'lft' => 11, 'rgt' => 12, 'depth' => 0]);

        Category::reguard();

        if (DB::connection()->getDriverName() === 'pgsql') {
            $tablePrefix  = DB::connection()->getTablePrefix();
            $sequenceName = $tablePrefix . 'categories_id_seq';
            DB::connection()->statement('ALTER SEQUENCE ' . $sequenceName . ' RESTART WITH 7');
        }
    }

    public function nestUptoAt($node, $levels = 10, $attrs = [])
    {
        Category::unguard();

        for ($i = 0; $i < $levels; $i++, $node = $new) {
            $new = Category::create(array_merge($attrs, ['name' => "{$node->name}.1"]));
            $new->makeChildOf($node);
        }

        Category::reguard();
    }
}

class ScopedCategorySeeder
{
    public function run()
    {
        ScopedCategory::unguard();

        ScopedCategory::create(['id' => 1, 'company_id' => 1, 'name' => 'Root 1', 'lft' => 1, 'rgt' => 10, 'depth' => 0]);
        ScopedCategory::create(['id' => 2, 'company_id' => 1, 'name' => 'Child 1', 'lft' => 2, 'rgt' => 3, 'depth' => 1, 'parent_id' => 1]);
        ScopedCategory::create(['id' => 3, 'company_id' => 1, 'name' => 'Child 2', 'lft' => 4, 'rgt' => 7, 'depth' => 1, 'parent_id' => 1]);
        ScopedCategory::create(['id' => 4, 'company_id' => 1, 'name' => 'Child 2.1', 'lft' => 5, 'rgt' => 6, 'depth' => 2, 'parent_id' => 3]);
        ScopedCategory::create(['id' => 5, 'company_id' => 1, 'name' => 'Child 3', 'lft' => 8, 'rgt' => 9, 'depth' => 1, 'parent_id' => 1]);

        ScopedCategory::create(['id' => 6, 'company_id' => 2, 'name' => 'Root 2', 'lft' => 1, 'rgt' => 10, 'depth' => 0]);
        ScopedCategory::create(['id' => 7, 'company_id' => 2, 'name' => 'Child 4', 'lft' => 2, 'rgt' => 3, 'depth' => 1, 'parent_id' => 6]);
        ScopedCategory::create(['id' => 8, 'company_id' => 2, 'name' => 'Child 5', 'lft' => 4, 'rgt' => 7, 'depth' => 1, 'parent_id' => 6]);
        ScopedCategory::create(['id' => 9, 'company_id' => 2, 'name' => 'Child 5.1', 'lft' => 5, 'rgt' => 6, 'depth' => 2, 'parent_id' => 8]);
        ScopedCategory::create(['id' => 10, 'company_id' => 2, 'name' => 'Child 6', 'lft' => 8, 'rgt' => 9, 'depth' => 1, 'parent_id' => 6]);

        ScopedCategory::reguard();

        if (DB::connection()->getDriverName() === 'pgsql') {
            $tablePrefix  = DB::connection()->getTablePrefix();
            $sequenceName = $tablePrefix . 'categories_id_seq';
            DB::connection()->statement('ALTER SEQUENCE ' . $sequenceName . ' RESTART WITH 11');
        }
    }
}

class MultiScopedCategorySeeder
{
    public function run()
    {
        MultiScopedCategory::unguard();

        MultiScopedCategory::create(['id' => 1, 'company_id' => 1, 'language' => 'en', 'name' => 'Root 1', 'lft' => 1, 'rgt' => 10, 'depth' => 0]);
        MultiScopedCategory::create(['id' => 2, 'company_id' => 1, 'language' => 'en', 'name' => 'Child 1', 'lft' => 2, 'rgt' => 3, 'depth' => 1, 'parent_id' => 1]);
        MultiScopedCategory::create(['id' => 3, 'company_id' => 1, 'language' => 'en', 'name' => 'Child 2', 'lft' => 4, 'rgt' => 7, 'depth' => 1, 'parent_id' => 1]);
        MultiScopedCategory::create(['id' => 4, 'company_id' => 1, 'language' => 'en', 'name' => 'Child 2.1', 'lft' => 5, 'rgt' => 6, 'depth' => 2, 'parent_id' => 3]);
        MultiScopedCategory::create(['id' => 5, 'company_id' => 1, 'language' => 'en', 'name' => 'Child 3', 'lft' => 8, 'rgt' => 9, 'depth' => 1, 'parent_id' => 1]);
        MultiScopedCategory::create(['id' => 6, 'company_id' => 2, 'language' => 'en', 'name' => 'Root 2', 'lft' => 1, 'rgt' => 10, 'depth' => 0]);
        MultiScopedCategory::create(['id' => 7, 'company_id' => 2, 'language' => 'en', 'name' => 'Child 4', 'lft' => 2, 'rgt' => 3, 'depth' => 1, 'parent_id' => 6]);
        MultiScopedCategory::create(['id' => 8, 'company_id' => 2, 'language' => 'en', 'name' => 'Child 5', 'lft' => 4, 'rgt' => 7, 'depth' => 1, 'parent_id' => 6]);
        MultiScopedCategory::create(['id' => 9, 'company_id' => 2, 'language' => 'en', 'name' => 'Child 5.1', 'lft' => 5, 'rgt' => 6, 'depth' => 2, 'parent_id' => 8]);
        MultiScopedCategory::create(['id' => 10, 'company_id' => 2, 'language' => 'en', 'name' => 'Child 6', 'lft' => 8, 'rgt' => 9, 'depth' => 1, 'parent_id' => 6]);
        MultiScopedCategory::create(['id' => 11, 'company_id' => 3, 'language' => 'fr', 'name' => 'Racine 1', 'lft' => 1, 'rgt' => 10, 'depth' => 0]);
        MultiScopedCategory::create(['id' => 12, 'company_id' => 3, 'language' => 'fr', 'name' => 'Enfant 1', 'lft' => 2, 'rgt' => 3, 'depth' => 1, 'parent_id' => 11]);
        MultiScopedCategory::create(['id' => 13, 'company_id' => 3, 'language' => 'fr', 'name' => 'Enfant 2', 'lft' => 4, 'rgt' => 7, 'depth' => 1, 'parent_id' => 11]);
        MultiScopedCategory::create(['id' => 14, 'company_id' => 3, 'language' => 'fr', 'name' => 'Enfant 2.1', 'lft' => 5, 'rgt' => 6, 'depth' => 2, 'parent_id' => 13]);
        MultiScopedCategory::create(['id' => 15, 'company_id' => 3, 'language' => 'fr', 'name' => 'Enfant 3', 'lft' => 8, 'rgt' => 9, 'depth' => 1, 'parent_id' => 11]);
        MultiScopedCategory::create(['id' => 16, 'company_id' => 3, 'language' => 'es', 'name' => 'Raiz 1',    'lft' => 1, 'rgt' => 10, 'depth' => 0]);
        MultiScopedCategory::create(['id' => 17, 'company_id' => 3, 'language' => 'es', 'name' => 'Hijo 1',    'lft' => 2, 'rgt' => 3, 'depth' => 1, 'parent_id' => 16]);
        MultiScopedCategory::create(['id' => 18, 'company_id' => 3, 'language' => 'es', 'name' => 'Hijo 2',    'lft' => 4, 'rgt' => 7, 'depth' => 1, 'parent_id' => 16]);
        MultiScopedCategory::create(['id' => 19, 'company_id' => 3, 'language' => 'es', 'name' => 'Hijo 2.1',    'lft' => 5, 'rgt' => 6, 'depth' => 2, 'parent_id' => 18]);
        MultiScopedCategory::create(['id' => 20, 'company_id' => 3, 'language' => 'es', 'name' => 'Hijo 3',    'lft' => 8, 'rgt' => 9, 'depth' => 1, 'parent_id' => 16]);

        MultiScopedCategory::reguard();

        if (DB::connection()->getDriverName() === 'pgsql') {
            $tablePrefix  = DB::connection()->getTablePrefix();
            $sequenceName = $tablePrefix . 'categories_id_seq';
            DB::connection()->statement('ALTER SEQUENCE ' . $sequenceName . ' RESTART WITH 21');
        }
    }
}

class OrderedCategorySeeder
{
    public function run()
    {
        OrderedCategory::unguard();

        OrderedCategory::create(['id' => 1, 'name' => 'Root Z', 'lft' => 1, 'rgt' => 10, 'depth' => 0]);
        OrderedCategory::create(['id' => 2, 'name' => 'Child C', 'lft' => 2, 'rgt' => 3, 'depth' => 1, 'parent_id' => 1]);
        OrderedCategory::create(['id' => 3, 'name' => 'Child G', 'lft' => 4, 'rgt' => 7, 'depth' => 1, 'parent_id' => 1]);
        OrderedCategory::create(['id' => 4, 'name' => 'Child G.1', 'lft' => 5, 'rgt' => 6, 'depth' => 2, 'parent_id' => 3]);
        OrderedCategory::create(['id' => 5, 'name' => 'Child A', 'lft' => 8, 'rgt' => 9, 'depth' => 1, 'parent_id' => 1]);
        OrderedCategory::create(['id' => 6, 'name' => 'Root A', 'lft' => 11, 'rgt' => 12, 'depth' => 0]);

        OrderedCategory::reguard();

        if (DB::connection()->getDriverName() === 'pgsql') {
            $tablePrefix  = DB::connection()->getTablePrefix();
            $sequenceName = $tablePrefix . 'categories_id_seq';
            DB::connection()->statement('ALTER SEQUENCE ' . $sequenceName . ' RESTART WITH 7');
        }
    }
}

class OrderedScopedCategorySeeder
{
    public function run()
    {
        OrderedScopedCategory::unguard();

        OrderedScopedCategory::create(['id' => 1, 'company_id' => 1, 'name' => 'Root 1', 'lft' => 1, 'rgt' => 10, 'depth' => 0]);
        OrderedScopedCategory::create(['id' => 2, 'company_id' => 1, 'name' => 'Child 1', 'lft' => 2, 'rgt' => 3, 'depth' => 1, 'parent_id' => 1]);
        OrderedScopedCategory::create(['id' => 3, 'company_id' => 1, 'name' => 'Child 2', 'lft' => 4, 'rgt' => 7, 'depth' => 1, 'parent_id' => 1]);
        OrderedScopedCategory::create(['id' => 4, 'company_id' => 1, 'name' => 'Child 2.1', 'lft' => 5, 'rgt' => 6, 'depth' => 2, 'parent_id' => 3]);
        OrderedScopedCategory::create(['id' => 5, 'company_id' => 1, 'name' => 'Child 3', 'lft' => 8, 'rgt' => 9, 'depth' => 1, 'parent_id' => 1]);

        OrderedScopedCategory::create(['id' => 6, 'company_id' => 2, 'name' => 'Root 2', 'lft' => 1, 'rgt' => 10, 'depth' => 0]);
        OrderedScopedCategory::create(['id' => 7, 'company_id' => 2, 'name' => 'Child 4', 'lft' => 2, 'rgt' => 3, 'depth' => 1, 'parent_id' => 6]);
        OrderedScopedCategory::create(['id' => 8, 'company_id' => 2, 'name' => 'Child 5', 'lft' => 4, 'rgt' => 7, 'depth' => 1, 'parent_id' => 6]);
        OrderedScopedCategory::create(['id' => 9, 'company_id' => 2, 'name' => 'Child 5.1', 'lft' => 5, 'rgt' => 6, 'depth' => 2, 'parent_id' => 8]);
        OrderedScopedCategory::create(['id' => 10, 'company_id' => 2, 'name' => 'Child 6', 'lft' => 8, 'rgt' => 9, 'depth' => 1, 'parent_id' => 6]);

        OrderedScopedCategory::reguard();

        if (DB::connection()->getDriverName() === 'pgsql') {
            $tablePrefix  = DB::connection()->getTablePrefix();
            $sequenceName = $tablePrefix . 'categories_id_seq';
            DB::connection()->statement('ALTER SEQUENCE ' . $sequenceName . ' RESTART WITH 11');
        }
    }
}
